src/components/multiselect/Multiselect.tsx

```diff
@@ -0,0 +1,252 @@
+      updateState(updateExternalState?.key, newValue);
+      setSelectedItems(newValue);
+      return newValue;
+    }
+
+    // If new selection is "all" we make sure to deselect all other options
+    if (newSelected.value === "all") {
+      updateState(updateExternalState?.key, [newSelected]);
+      setSelectedItems([newSelected]);
+      return [newSelected];
+    }
+
+    updateState(updateExternalState?.key, [
+      ...allCurrentlySelected,
+      newSelected
+    ]);
+    setSelectedItems([...allCurrentlySelected, newSelected]);
+    return [...allCurrentlySelected, newSelected];
+  };
+
+  useClickAway(ref, () => {
+    setIsDropdownOpen(false);
+  });
+
+  useDeepCompareEffect(() => {
+    setSelectedItems(initialSelectedOptions);
+  }, [setSelectedItems, initialSelectedOptions]);
+
+  const { getToggleButtonProps, getMenuProps, highlightedIndex, getItemProps } =
+    useSelect({
+      isOpen: isDropdownOpen,
+      selectedItem: null,
+      items: allOptions,
+      stateReducer: (state, actionAndChanges) => {
+        const { changes, type } = actionAndChanges;
+        switch (type) {
+          case useSelect.stateChangeTypes.MenuKeyDownEnter:
+          case useSelect.stateChangeTypes.MenuKeyDownSpaceButton:
+          case useSelect.stateChangeTypes.ItemClick:
+            return {
+              ...changes,
+              isOpen: true // keep the menu open after selection.
+            };
+          default:
+            break;
+        }
+        return changes;
+      },
+      onStateChange: ({ type, selectedItem: newSelectedItem }) => {
+        switch (type) {
+          case useSelect.stateChangeTypes.MenuKeyDownSpaceButton:
+          case useSelect.stateChangeTypes.ItemClick:
+          case useSelect.stateChangeTypes.MenuKeyDownEnter:
+            // If new selection isn't already selected, we add it
+            if (
+              newSelectedItem &&
+              !selectedItems.find(
+                (item) => item.value === newSelectedItem.value
+              )
+            ) {
+              setSelectedItems(
+                addNewSelectedItem(
+                  selectedItems,
+                  newSelectedItem,
+                  allOptions.length
+                )
+              );
+              return;
+            }
+            // If new selection is already selected, we deselect it
+            if (
+              newSelectedItem &&
+              selectedItems.find((item) => item.value === newSelectedItem.value)
+            ) {
+              // Unless it's the only selected item
+              if (selectedItems.length === 1) {
+                return;
+              }
+              const newSelectedItems = selectedItems.filter((item) => {
+                return item.value !== newSelectedItem.value;
+              });
+              setSelectedItems(newSelectedItems);
+            }
+            break;
+          default:
+            break;
+        }
+      }
+    });
```

Comment:
Is it possible to put this in a separate helper function. I could be a way to get a better overview of the multiselect component